hasher.go

```diff
@@ -30,22 +32,66 @@ func newHasher(hasherProvider IonHasherProvider) *hasher {
 	return &hasher{hasherProvider, currentHasher, hasherStack}
 }
 
-func (h *hasher) scalar() {
-	panic("implement me")
+func (h *hasher) scalar(ionValue hashValue) {
+	h.currentHasher.scalar(ionValue)
 }
 
 func (h *hasher) stepIn(ionValue hashValue) {
-	panic("implement me")
+	var hashFunction IonHasher
+
+	_, isStructSerializer := h.currentHasher.(structSerializer)
+	if isStructSerializer {
+		hashFunction = h.hasherProvider.newHasher()
+	} else {
+		hashFunction = h.currentHasher.(scalarSerializer).hashFunction
+	}
+
+	if ionValue.ionType() == ion.StructType {
+		h.currentHasher = newStructSerializer(hashFunction, 0, h.hasherProvider)
+	} else {
+		h.currentHasher = newScalarSerializer(hashFunction, 0)
+	}
+
+	h.hasherStack.push(h.currentHasher)
+	h.currentHasher.stepIn(ionValue)
 }
 
-func (h *hasher) stepOut() {
-	panic("implement me")
+func (h *hasher) stepOut() error {
+	if h.depth() == 0 {
+		return &InvalidOperationError{"hasher", "stepOut", "Hasher cannot step out any further"}
```

Comment:
I woiuld add to the message that depth is 0. That is the reason for not being able to step out and we should.

Reply:
Done.